app/models/manageiq/providers/amazon/agent_coordinator.rb

```diff
@@ -137,11 +137,28 @@ def deploy_agent
     @deploying = true
 
     kp = find_or_create_keypair
-    zone_name = ec2.client.describe_availability_zones.availability_zones.first.zone_name
-    subnets = get_subnets(zone_name)
-    raise "No subnet_id is available for #{zone_name}!" if subnets.empty?
-    subnet = subnets.first
-    security_group_id = find_or_create_security_group(subnet.vpc_id)
+    vpcs = get_dns_enabled_vpcs
+    raise "Smartstate analysis needs a VPC whose enableDnsSupport/enableDnsHostnames settings are true!" if vpcs.empty?
+
+    zone_names = ec2.client.describe_availability_zones.availability_zones.map(&:zone_name)
+    subnet = nil
+    zone_names.each do |zone_name|
+      vpcs.each do |vpc|
+        subnet = get_subnets(zone_name, vpc.vpc_id).try(:first)
+        break if subnet
+      end
+      break if subnet
```

Comment:
Same here...
Actually, this logic looks like it should probably be its own method.
```ruby
private def something_about_finding_the_subnet
  ec2.client.describe_availability_zones.availability_zones.detect do |availability_zone|
    vpcs.detect { get_subnets(availability_zone.zone_name, vpc.vpc_id).try(:first) }
  end || raise("No subnet is qualiified to deploy smartstate agent!")
end
```

Reply:
No, such a method will return us either vpc or availability_zone. What we need is subnet.

Reply:
`detect` fails me every time 😢 but `return` would work.

```ruby
private def something_about_finding_the_subnet
  ec2.client.describe_availability_zones.availability_zones.each do |availability_zone|
    vpcs.each do
      subnet = get_subnets(availability_zone.zone_name, vpc.vpc_id).try(:first)
      return subnet if subnet
    end
  end
  raise("No subnet is qualified to deploy smartstate agent!")
end
```

Reply:
Agreed! Will change as this, thanks!